File: design_patterns_udemy/01_SOLID/OCP.cpp
/*
 * From the Udemy course "Design Patterns in Modern C++"
 * https://www.udemy.com/course/patterns-cplusplus/
 * by Dmitri Nesteruk.
 * 
 * Open Closed Principle
 * Open for extension, closed for modification
 * 
 */

#include <string>
#include <vector>
#include <iostream>
using namespace std;

enum class Color { red, green, blue };
enum class Size { small, medium, large };

struct Product
{
  string name;
  Color color;
  Size size;
};

// Bad pattern: this product filter needs to be
// modified/extendedin ternally every time we add
// a new filter use case!
// Instead, we should follow the Open-Closed Principle:
// Open for extension, closed for modification.
// That can be achieved by defining pure virtual class templates
// that are inherited to create concrete filters.
struct ProductFilter
{
  typedef vector<Product*> Items;

  Items by_color(Items items, const Color color)
  {
    Items result;
    for (auto& i : items)
      if (i->color == color)
        result.push_back(i);
    return result;
  }

  Items by_size(Items items, const Size size)
  {
    Items result;
    for (auto& i : items)
      if (i->size == size)
        result.push_back(i);
    return result;
  }

  Items by_size_and_color(Items items, const Size size, const Color color)
  {
    Items result;
    for (auto& i : items)
      if (i->size == size && i->color == color)
        result.push_back(i);
    return result;
  }
};

template <typename T> struct AndSpecification;

// Good pattern: Open-Closed Principle:
// - open to extend by inheritance
// - closed to be modified in its definition
// We create virtual classes Specification & Filter
// and they are inherited to be concrete filter(s)
// and filtering specifications/conditions
template <typename T> struct Specification
{
  virtual ~Specification() = default;
  virtual bool is_satisfied(T* item) const = 0;

  // AndSpecification contains two specifications
  // but it breaks OCP if you add it post-hoc
  // thus, we need to define it outside
  /*AndSpecification<T> operator&&(Specification<T>&& other)
  {
    return AndSpecification<T>(*this, other);
  }*/
};

// AndSpecification contains two specifications.
// In order to make it more compact when using a pair of specifications
// we can define the && operator, which would return an AndSpecification
// from two single Specifications
template <typename T> AndSpecification<T> operator&&
  (const Specification<T>& first, const Specification<T>& second)
{
  return { first, second };
}

template <typename T> struct Filter
{
  virtual vector<T*> filter(vector<T*> items,
                            Specification<T>& spec) = 0;
};

struct BetterFilter : Filter<Product>
{
  vector<Product*> filter(vector<Product*> items,
                           Specification<Product> &spec) override
  {
    vector<Product*> result;
    for (auto& p : items)
      if (spec.is_satisfied(p))
        result.push_back(p);
    return result;
  }
};

struct ColorSpecification : Specification<Product>
{
  Color color;

  ColorSpecification(Color color) : color(color) {}

  bool is_satisfied(Product *item) const override {
    return item->color == color;
  }
};

struct SizeSpecification : Specification<Product>
{
  Size size;

  explicit SizeSpecification(const Size size)
    : size{ size }
  {
  }


  bool is_satisfied(Product* item) const override {
    return item->size == size;
  }
};

// AndSpecification contains two generic specifications,
// both are checked
template <typename T> struct AndSpecification : Specification<T>
{
  const Specification<T>& first;
  const Specification<T>& second;

  AndSpecification(const Specification<T>& first, const Specification<T>& second) 
    : first(first), second(second) {}

  bool is_satisfied(T *item) const override {
    return first.is_satisfied(item) && second.is_satisfied(item);
  }
};

int main()
{
  // Products
  Product apple{"Apple", Color::green, Size::small};
  Product tree{"Tree", Color::green, Size::large};
  Product house{"House", Color::blue, Size::large};

  // Vector of product pointers
  const vector<Product*> all { &apple, &tree, &house };

  // Filter which follows the open-close principle
  // with Filter and Specification
  BetterFilter bf;
  ColorSpecification green(Color::green);
  auto green_things = bf.filter(all, green);
  for (auto& x : green_things)
    cout << x->name << " is green\n";

  // Two specifications
  SizeSpecification large(Size::large);
  AndSpecification<Product> green_and_large(green, large);

  //auto big_green_things = bf.filter(all, green_and_large);

  // use the operator instead (same for || etc.)
  auto spec = green && large;
  for (auto& x : bf.filter(all, spec))
    cout << x->name << " is green and large\n";

  // Warning: the following will compile but will NOT work
  // auto spec2 = SizeSpecification{Size::large} &&
  //              ColorSpecification{Color::blue};

  //getchar();
  return 0;
}

File: design_patterns_udemy/01_SOLID/ocp.py
from enum import Enum

# The Enum base class
# makes possible to have enumerations
# like in C/C++
class Color(Enum):
    RED = 1
    GREEN = 2
    BLUE = 3


class Size(Enum):
    SMALL = 1
    MEDIUM = 2
    LARGE = 3

# Our product
class Product:
    def __init__(self, name, color, size):
        self.name = name
        self.color = color
        self.size = size

# We use generators with yield.
# yield suspends a function’s execution
# and sends a value back to the caller,
# but retains enough state to enable the function
# to resume where it left off.
# When the function resumes, 
# it continues execution immediately after the last yield run.
# This allows its code to produce a series of values over time, 
# rather than computing them at once and sending them back like a list.
# Commonly, yield is used within a loop.
class ProductFilter:
    def filter_by_color(self, products, color):
        for p in products:
            if p.color == color:
                yield p

    def filter_by_size(self, products, size):
        for p in products:
            if p.size == size:
                yield p

    def filter_by_size_and_color(self, products, size, color):
        for p in products:
            if p.color == color and p.size == size:
                yield p
    
    # BUT, this filter violates the OCP (principle)
    # OCP = open for extension, closed for modification
    # - once a class is written, we should not modify it
    # - but we should be able to extend it
    # This violation causes here additionally
    # a "state space explosion": as we add more properties,
    # the filter compbinations explode:
    # 2 criteria -> 3 filters/methods
    # 3 criteria (e.g., +Weight) -> c s w cs sw cw csw = 7 methods


# To tolve the issue of breaking OCP,
# we use a Specification + Filter classes.
# We're going to inherit override these classes,
# 
class Specification:
    def is_satisfied(self, item):
        pass

    # We overload the & operator, which makes life easier
    # In Python, we cannot overload the and operator,
    # but we can overload the & operator with __and__
    def __and__(self, other):
        return AndSpecification(self, other)

class Filter:
    def filter(self, items, spec):
        pass

# Color specification checker
class ColorSpecification(Specification):
    def __init__(self, color):
        self.color = color

    def is_satisfied(self, item):
        return item.color == self.color

# Size specification checker
class SizeSpecification(Specification):
    def __init__(self, size):
        self.size = size

    def is_satisfied(self, item):
        return item.size == self.size

# class AndSpecification(Specification):
#     def __init__(self, spec1, spec2):
#         self.spec2 = spec2
#         self.spec1 = spec1
#
#     def is_satisfied(self, item):
#         return self.spec1.is_satisfied(item) and \
#                self.spec2.is_satisfied(item)

# This is a COMBINATOR
# We give a variable number of arguments, i.e., specifications: *args
class AndSpecification(Specification):
    def __init__(self, *args):
        self.args = args

    def is_satisfied(self, item):
        # all(iterable) returns True if all items in an iterable are true,
        # otherwise it returns False.
        # If the iterable object is empty, the all() function also returns True.
        # map(fun, iterable): fun is applied to all elements of iterable
        # and the resulting iterable is returned
        # So, for all speficications in self.args
        # their is_satisfied(item) function is called.
        return all(map(lambda spec: spec.is_satisfied(item),
                       self.args))

# We could put this implementation in the base Filter class
# but instead, we do it here, because that enables
# inheriting Filter for other kind of Filter implementations
# -- that's also the OCP!
class BetterFilter(Filter):
    def filter(self, items, spec):
        for item in items:
            if spec.is_satisfied(item):
                yield item

if __name__ == "__main__":

    apple = Product('Apple', Color.GREEN, Size.SMALL)
    tree = Product('Tree', Color.GREEN, Size.LARGE)
    house = Product('House', Color.BLUE, Size.LARGE)

    products = [apple, tree, house]

    # BAD - Breaks OCP
    pf = ProductFilter()
    print('Green products (old):')
    for p in pf.filter_by_color(products, Color.GREEN):
        print(f' - {p.name} is green')

    # GOOD - OCP Compliant
    bf = BetterFilter()

    print('Green products (new):')
    green = ColorSpecification(Color.GREEN)
    for p in bf.filter(products, green):
        print(f' - {p.name} is green')

    print('Large products:')
    large = SizeSpecification(Size.LARGE)
    for p in bf.filter(products, large):
        print(f' - {p.name} is large')

    print('Large blue items:')
    # We have 2 options:
    # 1. We use AndSpecification
    # 2. We use the overloaded & operator from Specification, wehich returns AndSpecification
    # Both work!
    # 1. AndSpecification
    # large_blue = AndSpecification(large, ColorSpecification(Color.BLUE))
    # 2. Overloaded &
    large_blue = large & ColorSpecification(Color.BLUE)
    for p in bf.filter(products, large_blue):
        print(f' - {p.name} is large and blue')